### ISSUE TITLE
Base branch not displayed for resolving conflicts when dialog opens mid-rebase

### ISSUE THREAD

**Issue by kuychaco** [desktop/desktop#8113]
## Description

@tierninho originally discovered this bug in https://github.com/desktop/desktop/pull/8097#pullrequestreview-275217992. 

### The problem
There are a few ways we get to the points of presenting the `ShowConflictedFilesDialog` to users. 

1. The user initiates a rebase through the UI 
2. The user loads Desktop and there is already a rebase in progress
3. The user closed the `ShowConflictedFilesDialog` and then re-opened it by clicking "View Conflicts" in the `RebaseConflictsBanner`

For cases 2 and 3 the user will be disappointed to see that the name of the branch that they are rebasing onto (the base branch) is not present in the menu for resolving conflicts: 

![Screen Shot 2019-08-14 at 2 25 16 PM](https://user-images.githubusercontent.com/14828183/63065363-76950300-bea1-11e9-9b8f-acfc58832afa.png)

For case 1 (initiating a rebase through the UI) we pass along the `baseBranch` data to the `ShowConflictedFilesDialog` since we have it handy. 

It seems that for cases 2 and 3 we rely on `getRebaseInternalState` to provide the branch details to us. This method reads from the `.git/rebase-apply` folder which is what Git uses to keep track of where the user is at in a rebase operation. This folder contains only the commit sha for the branch being rebased onto, and not the branch name. Presumably, this is because it doesn't actually need the branch name in order to successfully complete a rebase... it just needs to know what commit to apply rebased commits onto. 

So we end up with `undefined` for the `baseBranch` and hence the missing branch name in the conflict resolution menu. 

### The solution?
How can we get access to the base branch name? 

**We could store off the base branch when the user selects it for the rebase.**
This would address case 3, but would not address case 2. Plus, where would we store this state? In memory? That doesn't help us if the user closes and re-opens Desktop. Do we store it somewhere in the `.git` folder on disk for future access? That doesn't help us if the user initiated the rebase on the command line, and generally feels a bit icky. 

**We could look through the branch reflog to see the most recent entry that says `rebase: checkout <base-branch>`**
This seems like it's the most reliable way to get the base branch, and I can't think of any edge cases it doesn't address. It will require adding another reflog method for shelling out to Git. 

## Version

* GitHub Desktop: bfd0207945 (dev)

* Operating system: macOS High Sierra 10.13.6

## Steps to Reproduce

1. download [repo-binary-file.zip](https://github.com/desktop/desktop/files/3494892/repo-binary-file.zip)
1. check out `master` and verify that you're on commit `bbb019c` (or `reset --hard` to it)
1. check out `branch` and verify that you're on commit `0359ae5` (or `reset --hard` to it)
1. on the command line in the repo type `git rebase master`. Verify that you see `CONFLICT (content): Merge conflict in cute_rodent.jpeg`
1. open Desktop (or close and then re-open it, or re-load the window if in dev mode)
1. click the "Resolve" button to see menu items for resolving the conflict

### Expected Behavior

`Use the modified file from master` is displayed as an option

### Actual Behavior

`Use the modified file` is displayed as an option (base branch name is missing)
<img width="730" alt="Fullscreen_8_15_19__9_20_PM" src="https://user-images.githubusercontent.com/7910250/63143330-995b1080-bfa2-11e9-9e65-80fc2e86d09d.png">

Alternative repro steps are to initiate a rebase via the UI, close the conflict dialogue, and then re-open it by clicking "View Conflicts" from the rebase message near the top of the window. However in my opinion this way of reproducing the bug doesn't catch as many edge cases as the one outlined above.

## Additional Information

Think the above pretty much covers it...

### Logs

N/A

**Comment by azad-derakhshani-GS** [comment]
Any news on this? The issue is still present in Github desktop 2.9.12 and very confusing.
I see the same two options as in @kuychaco's screenshot above, and the naming is totally misleading.

Until seeing this ticket here, I didn't even know which one of the two options does what, since they both say "modified file".
It should not say "modified" at all when it's referring to the target branch, since obviously there hasn't been any modification there.
The feature branch is the only one that modifies something.

So the naming should be changed to

- `Use the file from [target branch]`
- `Use the modified file from [feature branch]`

If the fix is somehow too difficult to implement, imho it would also be acceptable to hardcode it to:

- `Use the file from the target branch`
- `Use the modified file from your feature branch`

Or you could even do it like GitLab and call it `theirs` and `ours`, although personally I'm not a fan of that naming. But it still would be an improvement over the current naming.

**Comment by steveward** [comment]
Thanks for bumping this up @azad-derakhshani-GS, I was able to reproduce this issue as well and will mark it as a bug for us to look into.

**Comment by azad-derakhshani-GS** [comment]
Thats awesome, thank you!

**Comment by kokoshneta** [comment]
I’ve always been confused by this naming as well and never know which one to choose.

I realise the title of this issue specifies “mid-rebase”, but I see it a lot when pushing or pulling a branch from/to my local machine to/from a remote repo as well, and I assume the actual dialog is the same. But in the case of push or pull (correct me if I’m wrong) giving the name of the source and target branches alone would be even worse than the current implementation: since I’m usually pushing or pulling from `main` to `main`, the two options would have exactly the same name.

‘Feature branch’ doesn’t really work well in that case either, since the branches are the same, and though ‘target branch’ is an accurate description of the branch being rebased/pushed into/pulled into, which branch is the target will be different for different actions (e.g., when pushing, the remote branch is the target; when pulling, the local branch is).

But in all the cases I’ve come across the dialog, I mentally have a **local** version and a **remote** version of the file, and these are always the same:

- local = currently active branch on the computer I’m working on, being pushed from/pulled to/rebased
- remote = the branch being pulled from/pushed to/rebased from

This is essentially the same distinction as ‘ours’ and ‘theirs’, just without the awkwardness of those terms.

Are there cases where simply writing “Use the file from the local branch” and “Use the file from the remote branch” would not work?

**Comment by azad-resolver** [comment]
@kokoshneta Yeah I agree, calling them _local_ and _remote_ would be much better! 
Even if the wording might not always be accurate, it will always be clear to the user what is meant.
And I can't think of any cases where it wouldn't be helpful either.

**Comment by 0xdevalias** [comment]
> This is essentially the same distinction as ‘ours’ and ‘theirs’, just without the awkwardness of those terms.
> 
> Are there cases where simply writing “Use the file from the local branch” and “Use the file from the remote branch” would not work?

I've also never liked the 'ours'/'theirs' or similar wording that comes up in diff GUI's and similar cases; but don't really have much of a better solution.

Correct me if I'm wrong though, but at least theoretically, could you not be rebasing a local branch A onto a local branch B? In which case, they are both technically 'local'?

**Comment by tidy-dev** [comment]
Recap of this issue history:
1. Original issue is about when the base branch is missing when rebase originated elsewhere.
- Suggested solution is to use reflog to attempt to determine the base branch. 

2. Even with branch name. Providing a descriptor may make these options clearer and could be potential quick way of addressing the original issue since at least the relationship between the two branches would be communicated.

> could you not be rebasing a local branch A onto a local branch B? In which case, they are both technically 'local'?

Yes. Local/remote adds it's own bit of confusion in that case.  Other ideas: ours/theirs (tho stated above, that can be confusing), from/to, base/current, other/current.   Other thoughts: If we can detect the rebase is a remote to local operation, we could provide local/remote. If it is local to local like the case of a local merge, rebase, or interact rebase operation, provide a more generic descriptor. 

Scenario impact notes:
- Local rebase, merge, squash merge. 
- Interactive rebase -> reorder, cherry-pick, and squash
- Updating from remote -> rebase and merge.
